src/dataflow/src/source/kafka.rs

```diff
@@ -866,95 +814,81 @@ fn downgrade_capability(
+    // we have seen all records <= offset. We keep track of the last "closed" timestamp in that partition
+    // in next_partition_ts
+    if let Some(entries) = timestamp_histories.borrow_mut().get_mut(id) {
+        // Iterate over each partition that we know about
+        for (pid, entries) in entries {
+            let pid = match pid {
+                PartitionId::Kafka(pid) => *pid,
+                _ => unreachable!(
+                    "timestamp.rs should never send messages with non-Kafka partitions \
                                    to Kafka sources."
-                    ),
-                };
-
-                // There is an entry for a partition for which we have no metadata. Must refresh before
-                // continuing
-                if pid >= dp_info.current_partition_count {
-                    // PIDs in Kafka are monotonically increasing and 0-indexed. Finding a PID of x means there
-                    // must be at least (x+1) partitions
-                    dp_info.set_needs_refresh(pid + 1);
-                    return;
-                }
+                ),
+            };
 
-                let last_offset = cp_info.partition_metadata.get(&pid).unwrap().offset;
+            dp_info.ensure_partition(cp_info, pid);
 
-                // Check whether timestamps can be closed on this partition
-                while let Some((partition_count, ts, offset)) = entries.front() {
-                    assert!(
+            let last_offset = cp_info.partition_metadata.get(&pid).unwrap().offset;
+
+            // Check whether timestamps can be closed on this partition
+            while let Some((partition_count, ts, offset)) = entries.front() {
+                assert!(
                         *offset >= cp_info.start_offset,
                         "Internal error! Timestamping offset went below start: {} < {}. Materialize will now crash.",
                         offset, cp_info.start_offset
                     );
 
-                    assert!(
-                        *ts > 0,
-                        "Internal error! Received a zero-timestamp. Materialize will crash now."
-                    );
+                assert!(
+                    *ts > 0,
+                    "Internal error! Received a zero-timestamp. Materialize will crash now."
+                );
 
-                    // This timestamp update was done with the expectation that there were more partitions
-                    // than we know about. We have to refresh metadata before continuing
-                    if *partition_count > cp_info.get_partition_count() {
-                        // New partition has been added, must refresh metadata before continuing
-                        dp_info.set_needs_refresh(*partition_count);
-                        return;
-                    }
+                // This timestamp update was done with the expectation that there were more partitions
+                // than we know about.
```

Comment:
Is it worth adding a comment that we call "partition_count-1" as partition ids are contiguous/monotically increasing and start from 0?

Reply:
Oops, yeah, sorry for losing that comment.